### ISSUE TITLE
C stack overflow when using await in a loop

### ISSUE THREAD

**Issue by dogagithub** [TypeScriptToLua/TypeScriptToLua#1499]
Hello,

I've encountered a "C stack overflow" error while running the following TypeScript code which gets transpiled to Lua using TypeScriptToLua:
`for (let i = 0; i < 40; i++) {
    await delay(0.1);
    print('times:', i);
}`
The function delay is an asynchronous function that returns a Promise that resolves after the specified delay. The print function logs the current iteration to the console.

The issue occurs when the loop reaches the 36th iteration (i=35), at which point the program crashes with a "C stack overflow" error. The expected behavior would be for the loop to successfully complete all 40 iterations without any issues.

I've tried debugging the issue on my end but haven't been able to find a workaround. I've also ensured that the issue is not due to any memory leaks, high CPU usage, or problems with the delay function or the Lua runtime.

I'm using TypeScriptToLua version 1.20.1 and running the resulting Lua code in environment Warcraft III

Any help on this issue would be greatly appreciated.

Thank you!

**Comment by Zamiell** [comment]
the code works fine:

https://typescripttolua.github.io/play#code/IYZwngdgxgBAZgV2gFwJYHsIwCYFMA2wYAFBAgLYBcMZ5ARrgE4CU1ACo+uaiLgDwA3dKmwA+GAG8AvgCgZoSLEQoMWbOgAy6dAAdizSTJjH46RjGL5cyGKhgBeGAAYA3LZh8YAFle2A1H4GEkYmocAA7sCoNniEJE4AdACMzC4hocY6jKgQyMQA5GjkuCCU+QA0tqnpMLKyMupauvppQA

warcraft 3 has issues, let me try to find the issue that talks about it

**Comment by dogagithub** [comment]
> the code works fine:
> 
> https://typescripttolua.github.io/play#code/IYZwngdgxgBAZgV2gFwJYHsIwCYFMA2wYAFBAgLYBcMZ5ARrgE4CU1ACo+uaiLgDwA3dKmwA+GAG8AvgCgZoSLEQoMWbOgAy6dAAdizSTJjH46RjGL5cyGKhgBeGAAYA3LZh8YAFle2A1H4GEkYmocAA7sCoNniEJE4AdACMzC4hocY6jKgQyMQA5GjkuCCU+QA0tqnpMLKyMupauvppQA
> 
> warcraft 3 has issues, let me try to find the issue that talks about it

Hello,

Thank you for your reply. I tried the example you provided and indeed it works fine with a small number of iterations. However, when I increased the iteration count to 200, the program unexpectedly stopped at the 194th iteration without any error message.

Here is the link to the modified code:

https://typescripttolua.github.io/play/#code/IYZwngdgxgBAZgV2gFwJYHsIwCYFMA2wYAFBAgLYBcMZ5ARrgE4CU1ACo+uaiLgDwA3dKmwA+GAG8AvgCgZoSLEQoMWbOgAy6dAAdizSTJjH46RjGL5cyGKhgBeGAAYA3LZh8YAJievbAan8DCSMTMOAAd2BUGzxCEicAOgBGZhdQsOMdRlQIZGIAcjRyXBBKAoAaWzSMmFlZGXUtXX10oA

I'm not sure why the execution stops at this point and I would appreciate your help in understanding this.

Thank you!

**Comment by Zamiell** [comment]
ok i found the issue about this: https://github.com/TypeScriptToLua/TypeScriptToLua/issues/1466

**Comment by Zamiell** [comment]
i'm not sure why the playground stops at 194, but without knowing more i would guess that you just have warcraft 3 issues relating to the stop-the-world behavior as detailed in #1466

**Comment by Perryvw** [comment]
I've been looking into this and my suspicion is that the stack overflow and the playground stopping the loop early are both the same problem. My early theory is that somehow our async awaiter mechanism is somewhere, somehow constructing a large chain of promise callbacks (ie `then(p1 => p1.then(p2 => p2.then(p3 => p3.then(p4 => p4.then(....`) that at some point overwhelm the runtime by overflowing some kind of stack limit that can differ per environment.

I don't quite understand how this is happening, as I would expect the delay promise to 'break the chain'. So I could still be wrong and it could be something else. I will try to do some more investigation to figure out what is going on and if we can solve it.

If anyone else has bright ideas regarding what could be causing this please let me know.

**Comment by Perryvw** [comment]
Update: From adding some extra prints it seems almost certain were are running into some kind of stack limit. You can see the stack growing to large levels in this instrumented tstl-generated Lua: http://tpcg.io/_V9V3K7

Ending the last print before hanging with:
```
stack traceback:
	main.lua:266: in upvalue 'adopt'
	main.lua:287: in upvalue 'step'
	main.lua:274: in upvalue 'f'
	main.lua:232: in function <main.lua:230>
	[C]: in function 'pcall'
	main.lua:230: in local 'internalCallback'
	main.lua:154: in field 'then'
	main.lua:289: in upvalue 'step'
	main.lua:274: in upvalue 'f'
	main.lua:232: in function <main.lua:230>
...	(skipping 1163 levels)                                   <==== NOTE: stack is over 1100 levels deep!
	[C]: in function 'pcall'
	main.lua:230: in local 'internalCallback'
	main.lua:154: in field 'then'
	main.lua:289: in local 'step'
	main.lua:303: in upvalue 'executor'
	main.lua:111: in function <main.lua:110>
	[C]: in function 'pcall'
	main.lua:110: in method '____constructor'
	main.lua:42: in function <main.lua:40>
	(...tail calls...)
	main.lua:326: in main chunk
	[C]: in ?
```

So now the question becomes if we can somehow write this chain of coroutine resumes in a way that doesn't lead to the runtime treating it as a giant resursive call

**Comment by rhazarian** [comment]
Surprisingly, it seems that the problem is connected with `pcall`. I've read some random treads about problems people had with pcall + lua coroutines. Then, I've tried to remove `pcall` usage from your example: http://tpcg.io/_RP5AU8
With `pcall` is was crashing on 195th, however without it can successfully complete 3000 iterations, and possibly more.

**Comment by rhazarian** [comment]
I was able to narrow it down to the following place:

```
local ____try, ____hasReturned = pcall(function()
     self:handleCallbackData(
          f(nil, value),
          resolve,
          reject
     )
end)
```

I've rewritten it to the following:

```
local ____try, ____hasReturned = pcall(
    self.handleCallbackData,
    self, 
    f(nil, value),
    resolve,
    reject
)
```

And the issue is now gone: http://tpcg.io/_2NQ003

**Comment by rhazarian** [comment]
I'm not sure if it was the true root cause but not allocating this anonymous function seems to give unbelievably huge benefits.
While in its original form it dies on 195th iteration, its 'optimized' version was able to complete 50k iterations! I didn't test it any further.

I'll try to make a PR optimizing the Promise implementation.

We can also try to do more tail calls to optimize the stack depth.
For example: `resolve(nil, data)` at the end of the function is not a tail call and increases the stack depth.
`return resolve(nil, data)` is a tail call and doesn't increase the stack depth.

**Comment by rhazarian** [comment]
Updated: based on additional tests, the issue is not in the anonymous function but in calling `f` (`onFulfilled`/`onRejected`) from inside `pcall`. I have unintentionally omitted it from pcall block in my previous tests. However, if I add `pcall(f, nil, value)` the issue is back again. I'll think if we can avoid this pcall somehow.

**Comment by Perryvw** [comment]
Thanks for looking into this! I got stuck investigating this and couldn't find a good solution, hoping you can come up with a good solution.

I will summarize my own investigation because I don't think my previous comment was very clear:

When you do `for (X) { await Y }`, what you are really doing is creating a big chain of promises, one for each iteration of the loop. When one promise is resolved, the next one is triggered, and hte onresolve of that one triggers the next one, etc etc.

This means that the stack trace of the last promise points back to the stacktraces of all the promises before it, which somewhat obviously leads to hitting some stack limits. I have seen various comlaints about this but no real suggested solution. I think a possible way to fix this is if we could call some function without any preceding stacktrace, but I don't know how to do that.

Maybe there are other ways that I'm not aware of atm but I hit a bit of a dead end, hoping someone else has a good solution here!

**Comment by rhazarian** [comment]
> This means that the stack trace of the last promise points back to the stacktraces of all the promises before it, which somewhat obviously leads to hitting some stack limits. I have seen various comlaints about this but no real suggested solution. I think a possible way to fix this is if we could call some function without any preceding stacktrace, but I don't know how to do that.

It is true but Lua is quirky. Based on my tests, even if we remove `pcall` blocks the stack depth continues to grow each call no less than previously, however it doesn't crash. With `pcall` involved, the crash came when the stack depth was about 1000-1500 levels deep. However in my tests (without `pcall`s) it's able to function properly even with the stack depth up to 100k levels. It seems possible that Lua has some special optimizations for this case, but `pcall` explodes it.

Moreover, using `return`'s to trigger the tail call optimization seem to be extremely important as well.
For example, the two code snippets below are logically equivalent:

```
        if isPromiseLike(nil, data) then
            local nextpromise = data
            if nextpromise.state == 1 then
                resolve(nil, nextpromise.value)
            elseif nextpromise.state == 2 then
                reject(nil, nextpromise.rejectionReason)
            else
                data["then"](data, resolve, reject)
            end
        else
            resolve(nil, data)
        end
```

```
        if isPromiseLike(nil, data) then
            local nextpromise = data
            if nextpromise.state == 1 then
                return resolve(nil, nextpromise.value)
            elseif nextpromise.state == 2 then
                return reject(nil, nextpromise.rejectionReason)
            else
                return data["then"](data, resolve, reject)
            end
        else
            return resolve(nil, data)
        end
```

However, the first one ultimately causes Lua (not C!) stack overflow. The second one does not, because due to the tail call optimization the stack depth isn't increased.

**Comment by PyryM** [comment]
Yes the underlying issue is that there's a recursive generator pump loop that looks, simplified, like:
```lua
function pump_recursive(gen)
  local status, res = coroutine.resume(gen)
  if coroutine.status(gen) == "dead" then
    return status, res
  end
  return pump_recursive(gen)
end
```
which unrolls a loop inside a coroutine into a recursive call stack (note that this simplified pump implementation would work fine because it's tail recursive). The actual loop is spread between `Promise.then`, `Promise.createPromiseResolvingCallback`, and `Promise.handleCallbackData`, and with some effort could probably be made tail recursive.

However, I think ideally it wouldn't rely on tail call optimization which is likely to be brittle, but on an explicit iterative pump loop like:
```lua
function pump_iterative(gen)
  while true do
    local status, res = coroutine.resume(gen)
    if coroutine.status(gen) == "dead" then
        return status, res
    end
  end
end
```

**Comment by rhazarian** [comment]
> with some effort could probably be made tail recursive.

I've opened a PR that fixes the issue: https://github.com/TypeScriptToLua/TypeScriptToLua/pull/1530

Besides some other optimization, I've did exactly that!

> I think ideally it wouldn't rely on tail call optimization which is likely to be brittle, but on an explicit iterative pump loop like:

An explicit loop would be indeed better, however I don't see a way to adequately implement it keeping all of the JS Promise invariants. It's probably possible but would result in a very large and complex spaghetti code I guess.
